### PR TITLE
Fix xunit/xunit#1408 and xunit/xunit#1414 in xUnit2013 analyzer

### LINKED ISSUE #1408: AssertEqualShouldNotBeUsedForCollectionSizeCheck has same problem as #1406
Here: https://github.com/xunit/xunit.analyzers/blob/master/src/xunit.analyzers/AssertEqualShouldNotBeUsedForCollectionSizeCheck.cs#L92. The code is almost exactly the same as the code I just fixed for `Contains` in #1406. I don't have a repro yet, but I will find one & use it as a regression test when I get the time.

/cc @ErikSchierboom

### LINKED ISSUE #1414: Handle IReadOnlyCollection<T>.Count in xUnit2013
Spotted this in the Roslyn unit tests: https://github.com/dotnet/roslyn/pull/21335/files#diff-ebd4143e3b5ba1ccd2e25e728d9784bcR1867. It's calling `Assert.Equal(0, list.Count)` where `list` is a `SyntaxList<TNode>`. `SyntaxList<TNode>` only implements `IReadOnlyList<TNode>`, so xUnit2013 isn't firing.

/cc @ErikSchierboom

### PR DESCRIPTION
**Pull request by jamesqo**
Fixes xunit/xunit#1408, fixes xunit/xunit#1414.

/cc @marcind, @ErikSchierboom

### CODE REVIEW

**Review by ErikSchierboom — COMMENTED** [review]
As far as I can tell, it looks okay. I might be able to test it tomorrow.

**Comment by ErikSchierboom** [comment]
I have done some tests, and it appears to be working correctly.

File: src/xunit.analyzers/AssertEqualShouldNotBeUsedForCollectionSizeCheck.cs [hunk 1]
```diff
@@ -68,33 +69,38 @@ protected override void Analyze(SyntaxNodeAnalysisContext context, InvocationExp
...
+            var concreteICollectionOfT = containingType.GetGenericInterfaceImplementation(openICollectionOfT);
+            return concreteICollectionOfT != null && IsCountPropertyOf(concreteICollectionOfT, symbolInfo);
+        }
 
-            var genericCollectionCountSymbol = context.Compilation
-                .GetSpecialType(SpecialType.System_Collections_Generic_ICollection_T)
-                .Construct(symbolInfo.Symbol.ContainingType.TypeArguments.ToArray())
-                .GetMembers(nameof(ICollection<int>.Count))
-                .Single();
+        private static bool IsIReadOnlyCollectionOfTCountProperty(SyntaxNodeAnalysisContext context, SymbolInfo symbolInfo)
```
Comment (marcind) [inline src/xunit.analyzers/AssertEqualShouldNotBeUsedForCollectionSizeCheck.cs]: This method is almost identical to `IsICollectionOfTCountProperty`, other than a different value of the `SpecialType` enum. Could be refactored into a single implementation.

File: src/xunit.analyzers/AssertEqualShouldNotBeUsedForCollectionSizeCheck.cs [hunk 2]
```diff
@@ -32,8 +32,8 @@ public class AssertEqualShouldNotBeUsedForCollectionSizeCheck : AssertUsageAnaly
         protected override void Analyze(SyntaxNodeAnalysisContext context, InvocationExpressionSyntax invocation, IMethodSymbol method)
         {
             if (method.Parameters.Length != 2 ||
-                !method.Parameters[0].Type.SpecialType.Equals(SpecialType.System_Int32) ||
-                !method.Parameters[1].Type.SpecialType.Equals(SpecialType.System_Int32))
+                method.Parameters[0].Type.SpecialType != SpecialType.System_Int32 ||
```
Comment (marcind) [inline src/xunit.analyzers/AssertEqualShouldNotBeUsedForCollectionSizeCheck.cs]: Is this change necessary? I'd prefer to review as small a changeset as possible.

File: src/xunit.analyzers/AssertThrowsShouldNotBeUsedForAsyncThrowsCheckFixer.cs [hunk 3]
```diff
@@ -74,7 +74,7 @@ private static SyntaxTokenList GetModifiersWithAsyncKeywordAdded(MethodDeclarati
 
             var semanticModel = await document.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false);
             var methodSymbol = semanticModel.GetSymbolInfo(method.ReturnType, cancellationToken).Symbol as ITypeSymbol;
-            var taskType = semanticModel.Compilation.GetTypeByMetadataName(typeof(Task).FullName);
+            var taskType = semanticModel.Compilation.GetTypeByMetadataName(Constants.Types.SystemThreadingTasksTask);
```
Comment (marcind) [inline src/xunit.analyzers/AssertThrowsShouldNotBeUsedForAsyncThrowsCheckFixer.cs]: This file isn't even related to the bugs referenced by this commit.

File: src/xunit.analyzers/FixProviders/MemberFixBase.cs [hunk 4]
```diff
@@ -27,7 +27,7 @@ public MemberFixBase(IEnumerable<string> diagnostics)
             if (declaringType == null)
                 return;
 
-            var member = declaringType.GetMembers(diagnostic.Properties["MemberName"]).FirstOrDefault();
+            var member = declaringType.GetMember(diagnostic.Properties["MemberName"]);
```
Comment (marcind) [inline src/xunit.analyzers/FixProviders/MemberFixBase.cs]: Once again, not a bad change but I'd prefer to separate this sort of cleanup into different PRs/commits.

File: src/xunit.analyzers/AssertEqualShouldNotBeUsedForCollectionSizeCheck.cs (hunk 2 above)
Comment (jamesqo) [reply src/xunit.analyzers/AssertEqualShouldNotBeUsedForCollectionSizeCheck.cs]: Not strictly necessary, so I'll revert it.

File: src/xunit.analyzers/FixProviders/MemberFixBase.cs (hunk 4 above)
Comment (jamesqo) [reply src/xunit.analyzers/FixProviders/MemberFixBase.cs]: OK, I'll open a separate PR for this. You mentioned earlier you were interested more in feature/bug PRs than refactoring PRs, so I was trying to sneak in a bit of refactoring into the former kind of PR.

Comment (marcind) [reply src/xunit.analyzers/FixProviders/MemberFixBase.cs]: > trying to sneak in

being the operative expression here :).

Please consider that I have a limited amount of personal time I'm willing to spend on this project and so the less "noise" I have to filter through, the more I can spend on the "signal".

Comment (jamesqo) [reply src/xunit.analyzers/FixProviders/MemberFixBase.cs]: > > trying to sneak in
>
> being the operative expression here :).

I was half-joking there :smile: But anyway, here's what I was trying to get at in that comment: what kinds of refactoring PRs will you take in and what kinds will you reject? I'm just a bit confused because #72 was accepted and you comment here you'd like a different PR for the refactorings, but in https://github.com/xunit/xunit.analyzers/pull/73#issuecomment-322104265 you said

> I would much rather review and accept PRs that contain new features and bug fixes (as taken from the issue tracker) than refactorings and stylistic changes.

Comment (marcind) [reply src/xunit.analyzers/FixProviders/MemberFixBase.cs]: I like PRs to contain a single commit that references an issue (bug or new feature) from the issue tracker. I use the commit history to collect release notes and so would rather have commits that express a self-contained piece of functional change.

I recognize that sometimes a functional change is easier to implement after refactoring the existing code. In those cases I prefer the PR to have one refactoring commit and one commit that implements the functional change. I will preserve the 2 commits via a rebase.

I rewrite local git history all the time and do not like seeing commits like "WIP", "fix the test", or "forgot to change the thing in this one place" within a PR. I see no value is such commits being preserved in the project's history and think they should be squashed whenever possible. I have no qualms about force pushing a branch.

For refactorings I like to see a meaningful improvement to the reusability/readability of the code. Factoring out identical or similar pieces of code into a reusable function is great. Stylistic (e.g. bracing style, whitespace) or inconsequential structural (e.g. expression bodied vs. traditional functions/properties, referencing different constants, changing a static to a const, extensions vs regular static calls) are not great. Drive-by tweaks to a file unrelated to the functional change addressed in the PR are simply an unnecessary distraction.

TL;DR: PRs that add value (new features/bug fixes) get accepted. PRs that move my cheese around get rejected.

PS: I'm a fickle and capricious person and can arbitrarily misapply these rules at any time.

**Review by marcind — CHANGES_REQUESTED** [review]
Waiting for changes

**Comment by marcind** [comment]
@jamesqo Will you be able to get this fixed up sometime soon? I'd like to cut a new release within a day so that it can be included in the upcoming xUnit beta. If not, I can make the changes myself.

**Comment by jamesqo** [comment]
@marcind Sure, I'll make time to fix this today. You can expect me to respond to feedback w/I a couple of hours once I'm near a computer.

**Comment by jamesqo** [comment]
@marcind I've finished revising this PR, PTAL
